internal/provider/resource_credential.go

```diff
@@ -0,0 +1,277 @@
+package provider
+
+import (
+	"context"
+	"fmt"
+	"strconv"
+	"strings"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	"github.com/juju/terraform-provider-juju/internal/juju"
+)
+
+func resourceCredential() *schema.Resource {
+	return &schema.Resource{
+		Description: "A resource that represent a credential for a cloud.",
+
+		CreateContext: resourceCredentialCreate,
+		ReadContext:   resourceCredentialRead,
+		UpdateContext: resourceCredentialUpdate,
+		DeleteContext: resourceCredentialDelete,
+		Importer: &schema.ResourceImporter{
+			StateContext: resourceCredentialImporter,
+		},
+
+		Schema: map[string]*schema.Schema{
+			"cloud": {
+				Description: "JuJu Cloud where the credentials will be used to access",
+				Type:        schema.TypeList,
+				Optional:    true,
+				Computed:    true,
+				ForceNew:    true,
+				MaxItems:    1,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"name": &schema.Schema{
+							Description: "The name of the cloud",
+							Type:        schema.TypeString,
+							Required:    true,
+						},
+					},
+				},
+			},
+			"attributes": {
+				Description: "Credential attributes accordingly to the cloud",
+				Type:        schema.TypeMap,
+				Optional:    true,
+				Elem:        &schema.Schema{Type: schema.TypeString, Optional: true},
+			},
+			"auth_type": {
```

Comment:
It seems that the auth types vary accordingly to the cloud so to do this validation it would be necessary to call the API and then get the auth types available for the cloud specified. Something like how its done in the CLI:
https://github.com/juju/juju/blob/juju-2.9.37/cmd/juju/cloud/addcredential.go#L255

Do you think we should add it?

At least when adding to the controller, this is validated and you get the error:
```
Error: updating cloud credentials: validating credential "microk8s/admin/mycredential" for cloud "microk8s": supported auth-types ["certificate" "clientcertificate" "oauth2" "oauth2withcert" "userpass"], "banana" not supported
```

Reply:
I think so. We can simply have the complete list of supported `auth_type` and if this is not supported by the target cloud let the API to report the error. The other approach (and more interesting) is to have a ValidateFunction here to check if the auth type is valid for the target cloud.